src/message_expander.lua

```diff
@@ -0,0 +1,167 @@
+        end
+    end
+end
+
+local function is_any()
+    return true
+end
+
+local function is_opening_bracket(token)
+    return token == "{"
+end
+
+local function is_closing_bracket(token)
+    return token == "}"
+end
+
+local function is_pipe(token)
+
+    return token == "|"
+end
+
+local function is_u (token)
+    return token == "u"
+end
+
+local function is_not_bracket(token)
+    return not is_opening_bracket(token) and not is_closing_bracket(token)
+end
+
+local function add_token(self, token)
+    table.insert(self.tokens_, token)
+end
+
+local function add_open(self, token)
+    table.insert(self.tokens_, "{")
+    table.insert(self.tokens_, token)
+end
+
+local function add_parameter_name(self, token)
+    table.insert(self.last_parameter_.characters, token)
+end
+
+local function set_unquoted(self)
+    self.last_parameter_.quote = false
+end
+
+local function unwrap_parameter_value(parameter)
+    if parameter == nil then
+        return "missing value"
+    else
+        local type = type(parameter)
+        if (type == "table") then
+            return parameter.value
+        else
+            return parameter
+        end
+    end
+end
+
+local function replace_parameter(self)
+    local parameter_name = table.concat(self.last_parameter_.characters)
+    local value = unwrap_parameter_value(self.parameters[parameter_name])
+    local type = type(value)
+    if (type == "string") and (self.last_parameter_.quote) then
+        table.insert(self.tokens_, "'")
+        table.insert(self.tokens_, value)
+        table.insert(self.tokens_, "'")
+    else
+        table.insert(self.tokens_, value)
+    end
+    self.last_parameter_.characters = {}
+    self.last_parameter_.quote = true
+end
+
+local function replace_and_add(self, token)
+    replace_parameter(self)
+    add_token(self, token)
+end
+
+local function do_nothing() end
+
+M.transitions_ = {
+    {"TEXT"              , is_not_bracket    , add_token         , "TEXT"              },
+    {"TEXT"              , is_opening_bracket, do_nothing        , "OPEN_1"            },
+    {"OPEN_1"            , is_opening_bracket, do_nothing        , "PARAMETER"         },
+    {"OPEN_1"            , is_any            , add_open          , "TEXT"              },
+    {"PARAMETER"         , is_closing_bracket, do_nothing        , "CLOSE_1"           },
+    {"PARAMETER"         , is_pipe           , do_nothing        , "SWITCH"            },
+    {"PARAMETER"         , is_any            , add_parameter_name, "PARAMETER"         },
+    {"SWITCH"            , is_closing_bracket, do_nothing        , "CLOSE_1"           },
+    {"SWITCH"            , is_u              , set_unquoted      , "SWITCH"            },
+    {"SWITCH"            , is_any            , do_nothing        , "SWITCH"            },
+    {"CLOSE_1"           , is_closing_bracket, replace_parameter , "TEXT"              },
+    {"CLOSE_1"           , is_any            , replace_and_add   , "TEXT"              }
```

Comment:
This transition seems strange to me.
It would be called for example by `test: {{file}`.
This error message is however, invalid.
Also consider:
`{{test{1} }} `

Reply:
It's intentionally forgiving. You don't want to produce an error that an error was not formulated correctly. You want to get out as much as possible from the original message.

But of course you can only be fault tolerant for the obvious problems.